internal/entitlement/metered/grant_owner_adapter.go

```diff
@@ -121,50 +121,70 @@ func (e *entitlementGrantOwner) GetPeriodStartTimesBetween(ctx context.Context,
 	return times, nil
 }
 
-func (e *entitlementGrantOwner) EndCurrentUsagePeriod(ctx context.Context, owner credit.NamespacedGrantOwner, at time.Time) error {
-	// Check if time is after current start time. If so then we can end the period
-	currentStartAt, err := e.GetUsagePeriodStartAt(ctx, owner, at)
-	if err != nil {
-		return fmt.Errorf("failed to get current usage period start time: %w", err)
-	}
-	if at.Before(currentStartAt) || at.Equal(currentStartAt) {
-		return fmt.Errorf("can only end usage period after current period start time")
-	}
-
-	// Save usage reset
-	return e.usageResetRepo.Save(ctx, UsageResetTime{
-		NamespacedModel: models.NamespacedModel{
-			Namespace: owner.Namespace,
-		},
-		EntitlementID: owner.NamespacedID().ID,
-		ResetTime:     at,
-	})
-}
-
 // FIXME: this is a terrible hack, write generic Atomicity stuff for connectors...
-func (e *entitlementGrantOwner) EndCurrentUsagePeriodTx(ctx context.Context, tx *entutils.TxDriver, owner credit.NamespacedGrantOwner, at time.Time) error {
+func (e *entitlementGrantOwner) EndCurrentUsagePeriodTx(ctx context.Context, tx *entutils.TxDriver, owner credit.NamespacedGrantOwner, params credit.EndCurrentUsagePeriodParams) error {
 	_, err := entutils.RunInTransaction(ctx, tx, func(ctx context.Context, tx *entutils.TxDriver) (*interface{}, error) {
 		// Check if time is after current start time. If so then we can end the period
-		currentStartAt, err := e.GetUsagePeriodStartAt(ctx, owner, at)
+		currentStartAt, err := e.GetUsagePeriodStartAt(ctx, owner, params.At)
 		if err != nil {
 			return nil, fmt.Errorf("failed to get current usage period start time: %w", err)
 		}
-		if at.Before(currentStartAt) || at.Equal(currentStartAt) {
+		if !params.At.After(currentStartAt) {
 			return nil, &models.GenericUserError{Message: "can only end usage period after current period start time"}
 		}
 
+		if err := e.updateEntitlementUsagePeriod(ctx, tx, owner, params); err != nil {
+			return nil, fmt.Errorf("failed to update entitlement usage period: %w", err)
+		}
+
 		// Save usage reset
 		return nil, e.usageResetRepo.WithTx(ctx, tx).Save(ctx, UsageResetTime{
 			NamespacedModel: models.NamespacedModel{
 				Namespace: owner.Namespace,
 			},
 			EntitlementID: owner.NamespacedID().ID,
-			ResetTime:     at,
+			ResetTime:     params.At,
 		})
 	})
 	return err
 }
 
+func (e *entitlementGrantOwner) updateEntitlementUsagePeriod(ctx context.Context, tx *entutils.TxDriver, owner credit.NamespacedGrantOwner, params credit.EndCurrentUsagePeriodParams) error {
+	er := e.entitlementRepo.WithTx(ctx, tx)
+
+	entitlementEntity, err := er.GetEntitlement(ctx, owner.NamespacedID())
+	if err != nil {
+		return err
+	}
+
+	if entitlementEntity.UsagePeriod == nil {
+		e.logger.Warn("metered entitlement does not have a usage period set", "entitlement", owner.NamespacedID().String())
+		return nil
```

Comment:
I don't think this is the reasonable default behavior, it's just convenient for you here

It should attempt the update based on the inputs & fail if it's not possible. Also, it's arguments could include the entire usageperiod

Reply:
This should not happen, but agree, changed the behavior.